src/main/scala/com/fulcrumgenomics/util/SampleBarcodeMetric.scala

```diff
@@ -0,0 +1,112 @@
+  }
+
+  def finalizeMetrics(barcodeToMetrics: mutable.Map[String, SampleBarcodeMetric],
+                      noMatchBarcode: String): Unit = {
+    def updateBarcodeMetrics(barcode: String, metric: SampleBarcodeMetric): Unit = barcodeToMetrics(barcode) = metric
+    def noMatchMetric: SampleBarcodeMetric = barcodeToMetrics(noMatchBarcode)
+    
+    var totalReads: Int           = noMatchMetric.READS
+    var totalPfReads: Int         = noMatchMetric.PF_READS
+    var totalPfReadsAssigned: Int = 0
+
+    barcodeToMetrics.foreach { case (barcode, metric) =>
+      totalReads           += metric.READS
+      totalPfReads         += metric.PF_READS
+      totalPfReadsAssigned += metric.PF_READS
+    }
+
+    if (totalReads > 0) {
+      updateBarcodeMetrics(noMatchBarcode, noMatchMetric.copy(PCT_MATCHES = noMatchMetric.READS / totalReads.toDouble))
+      var bestPctOfAllBarcodeMatches: Double = 0
+      barcodeToMetrics.foreach { case (barcode, metric) =>
+        val pctMatches = metric.READS / totalReads.toDouble
+        if (pctMatches > bestPctOfAllBarcodeMatches) {
+          bestPctOfAllBarcodeMatches = pctMatches
+        }
+        updateBarcodeMetrics(barcode, metric.copy(PCT_MATCHES = pctMatches))
+      }
+      if (bestPctOfAllBarcodeMatches > 0) {
+        updateBarcodeMetrics(noMatchBarcode, noMatchMetric.copy(RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT = noMatchMetric.PCT_MATCHES / bestPctOfAllBarcodeMatches))
+        barcodeToMetrics.foreach { case (barcode, metric) =>
+          updateBarcodeMetrics(barcode, metric.copy(RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT = metric.PCT_MATCHES / bestPctOfAllBarcodeMatches))
+        }
+      }
+    }
+    if (totalPfReads > 0) {
+      var bestPctOfAllBarcodeMatches: Double = 0
+      barcodeToMetrics.foreach { case (barcode, metric) =>
+        val pctPfMatches = metric.PF_READS / totalPfReads.toDouble
+        if (pctPfMatches > bestPctOfAllBarcodeMatches) {
+          bestPctOfAllBarcodeMatches = pctPfMatches
+        }
+        updateBarcodeMetrics(barcode,  metric.copy(PF_PCT_MATCHES = pctPfMatches))
+      }
+      if (bestPctOfAllBarcodeMatches > 0) {
+        updateBarcodeMetrics(noMatchBarcode, noMatchMetric.copy(PF_RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT = noMatchMetric.PF_PCT_MATCHES / bestPctOfAllBarcodeMatches))
+        barcodeToMetrics.foreach { case (barcode, metric) =>
+          updateBarcodeMetrics(barcode,  metric.copy(PF_RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT = metric.PF_PCT_MATCHES / bestPctOfAllBarcodeMatches))
+        }
+      }
+
+    }
+    if (totalPfReadsAssigned > 0) {
+      val mean: Double = totalPfReadsAssigned.toDouble / barcodeToMetrics.values.size.toDouble
+      barcodeToMetrics.foreach { case (barcode, metric) =>
+        updateBarcodeMetrics(barcode,  metric.copy(PF_NORMALIZED_MATCHES = metric.PF_READS / mean))
+      }
+    }
+  }
+}
+
+/**
+  * Metrics for matching reads to sample barcodes.
+  */
+case class SampleBarcodeMetric(
```

Comment:
My temptation would be to go all-lower, either with underscores or with camel case.  Other than re-parsing programatically with scala, my primary use case would be to pull stuff into R or similar, and using all-caps names is (albeit a minor) annoyance.